File: lib/src/facts/windows/virtualization_resolver.cc
#include <internal/facts/windows/virtualization_resolver.hpp>
#include <leatherman/windows/wmi.hpp>
#include <facter/facts/scalar_value.hpp>
#include <facter/facts/collection.hpp>
#include <facter/facts/fact.hpp>
#include <facter/facts/vm.hpp>
#include <leatherman/logging/logging.hpp>
#include <vector>
#include <tuple>

using namespace std;
using namespace facter::facts;
using namespace leatherman::windows;

namespace facter { namespace facts { namespace windows {

    virtualization_resolver::virtualization_resolver(shared_ptr<wmi> wmi_conn) :
        resolvers::virtualization_resolver(),
        _wmi(move(wmi_conn))
    {
    }

    string virtualization_resolver::get_hypervisor(collection& facts)
    {
        static vector<tuple<string, string>> vms = {
            make_tuple("VirtualBox",        string(vm::virtualbox)),
            make_tuple("VMware",            string(vm::vmware)),
            make_tuple("KVM",               string(vm::kvm)),
            make_tuple("Bochs",             string(vm::bochs)),
            make_tuple("Google",            string(vm::gce)),
            make_tuple("OpenStack",         string(vm::openstack)),
        };

        auto vals = _wmi->query(wmi::computersystem, {wmi::manufacturer, wmi::model});
        if (vals.empty()) {
            return {};
        }

        auto &manufacturer = wmi::get(vals, wmi::manufacturer);
        auto &model = wmi::get(vals, wmi::model);

        for (auto const& vm : vms) {
            if (model.find(get<0>(vm)) != string::npos) {
                return get<1>(vm);
            }
        }

        if (model.find("Virtual Machine") != string::npos && manufacturer.find("Microsoft") != string::npos) {
            return vm::hyperv;
        }

        if (manufacturer.find("Xen") != string::npos) {
            return vm::xen;
        }

        if (manufacturer.find("Amazon EC2") != string::npos) {
            return vm::kvm;
        }

        return {};
    }

}}}  // namespace facter::facts::windows

File: lib/facter/resolvers/windows/virtualization_resolver.rb
# frozen_string_literal: true

module Facter
  module Resolvers
    class Virtualization < BaseResolver
      @log = Facter::Log.new(self)
      @semaphore = Mutex.new
      @fact_list ||= {}

      class << self
        # Virtual
        # Is_Virtual

        MODEL_HASH = { 'VirtualBox' => 'virtualbox', 'VMware' => 'vmware', 'KVM' => 'kvm',
                       'Bochs' => 'bochs', 'Google' => 'gce', 'OpenStack' => 'openstack' }.freeze

        private

        def post_resolve(fact_name)
          @fact_list.fetch(fact_name) { read_fact_from_computer_system(fact_name) }
        end

        def read_fact_from_computer_system(fact_name)
          win = Win32Ole.new
          comp = win.exec_query('SELECT Manufacturer,Model,OEMStringArray FROM Win32_ComputerSystem')
          unless comp
            @log.debug 'WMI query returned no results for Win32_ComputerSystem with values'\
            ' Manufacturer, Model and OEMStringArray.'
            return
          end

          build_fact_list(comp)
          @fact_list[fact_name]
        end

        def determine_hypervisor_by_model(comp)
          MODEL_HASH[MODEL_HASH.keys.find { |key| comp.Model =~ /^#{key}/ }]
        end

        def determine_hypervisor_by_manufacturer(comp)
          manufacturer = comp.Manufacturer
          if comp.Model =~ /^Virtual Machine/ && manufacturer =~ /^Microsoft/
            'hyperv'
          elsif manufacturer =~ /^Xen/
            'xen'
          elsif manufacturer =~ /^Amazon EC2/
            'kvm'
          else
            'physical'
          end
        end

        def build_fact_list(comp)
          @fact_list[:oem_strings] = []
          @fact_list[:oem_strings] += comp.to_enum.map(&:OEMStringArray).flatten

          comp = comp.to_enum.first
          hypervisor = determine_hypervisor_by_model(comp) || determine_hypervisor_by_manufacturer(comp)

          @fact_list[:virtual] = hypervisor
          @fact_list[:is_virtual] = hypervisor.include?('physical') ? false : true
        end
      end
    end
  end
end
